// @nuxt/hints/dist/runtime/third-party-scripts/plugin.client.js
import { defineNuxtPlugin, ref, useNuxtApp } from "#imports";
const EXTENSIONS_SCHEMES_RE = /^(chrome-extension|moz-extension|safari-extension|ms-browser-extension):/;
function isExtensionScript(src) {
  try {
    const url = new URL(src, window.location.origin);
    return EXTENSIONS_SCHEMES_RE.test(url.protocol);
  } catch {
    return false;
  }
}
function isSameOriginScript(src) {
  try {
    const url = new URL(src, window.location.origin);
    return url.origin === window.location.origin;
  } catch {
    return false;
  }
}
function isIgnoredScript(src) {
  return isSameOriginScript(src) || isExtensionScript(src);
}
export default defineNuxtPlugin({
  name: "nuxt-hints:third-party-scripts",
  setup() {
    const nuxtApp = useNuxtApp();
    const scripts = nuxtApp.__hints_tpc = ref([]);
    const isUsingNuxtScripts = !!nuxtApp.$scripts;
    nuxtApp.hook("hints:scripts:added", (script) => {
      scripts.value.push({ element: script, loaded: false });
    });
    nuxtApp.hook("hints:scripts:loaded", (script) => {
      const existingScript = scripts.value.find((s) => s.element === script);
      if (existingScript) {
        existingScript.loaded = true;
      } else {
        console.warn(`[@nuxt/hints]: Script loaded event received for a script not tracked: ${script.src}. Please open an issue with a minimal reproduction if you think this is a bug.`);
        scripts.value.push({ element: script, loaded: true });
      }
    });
    nuxtApp.hooks.hookOnce("app:mounted", () => {
      let hasThirdPartyScript = false;
      for (const script of document.scripts) {
        if (script.src && !isIgnoredScript(script.src)) {
          hasThirdPartyScript = true;
          onScriptAdded(script);
        }
      }
      if (hasThirdPartyScript && !isUsingNuxtScripts) {
        console.info("\u2139\uFE0F [@nuxt/hints]: Third-party scripts detected on page load: consider using @nuxt/scripts");
      }
    });
    const observer = new MutationObserver((mutations) => {
      for (const mutation of mutations) {
        if (mutation.type === "childList") {
          for (const node of mutation.addedNodes) {
            if (isScript(node) && node.src && !isIgnoredScript(node.src)) {
              onScriptAdded(node);
            }
          }
        }
      }
    });
    observer.observe(document.documentElement, {
      childList: true,
      subtree: true
    });
    function onScriptAdded(script) {
      if (!script.crossOrigin) {
        console.warn(`[@nuxt/hints]: Third-party script "${script.src}" is missing crossorigin attribute. Consider adding crossorigin="anonymous" for better security and error reporting.`);
      }
      nuxtApp.callHook("hints:scripts:added", script).then(() => {
        if (!script.loaded) {
          script.addEventListener("load", () => {
            window.__hints_TPC_saveTime(script, script.__hints_TPC_start_time);
            nuxtApp.callHook("hints:scripts:loaded", script);
          });
        } else {
          window.__hints_TPC_saveTime(script, script.__hints_TPC_start_time);
          nuxtApp.callHook("hints:scripts:loaded", script);
        }
      });
      console.info(`\u2139\uFE0F [@nuxt/hints]: Dynamically added third-party script detected: ${script.src}`);
    }
  }
});
function isScript(node) {
  return node.nodeName === "SCRIPT";
}
